README.md
# TV Maniac
### Python desktop application connecting via API to https://www.themoviedb.org/ (GUI made in tkinter)

### Functionalities:
- getting data from a configuration file
- displaying the most popular movies and TV series of the previous year
- searching for movies and series using filtering and sorting
- displaying details of each production
- keeping and displaying history of recently viewed productions
- switching the incognito mode, in which the browsing history is not saved
- clearing browsing history

### Installation and launching:
1. Install required modules:
```
pip install requests
pip install pillow
```
2. Edit the configuration file (config_file.txt) as needed
3. Start application
```
py main_app.py
```
4. Start view schould look like this:\
![Start view image](https://github.com/klaudial99/tv-maniac/raw/master/images/start-view.jpg)

#### Remark
The application was supposed to perform all operations on data sets itself, so it uses only the basic API features. Hence, the application does not run fast.\
During the first run on a given day, configuration files are updated, thanks to which the application loads faster on subsequent launchings.


history.txt
tv 102903
tv 100834
tv 98986
tv 96160
movie 454626
movie 508442
tv 87784
tv 87739


scroll.py
import tkinter as tk


class ScrollableFrame(tk.Frame):

    """
    creates frame with scroll
    """

    def __init__(self, container, *args, **kwargs):
        super().__init__(container, *args, **kwargs)
        self.canvas = tk.Canvas(self)
        self.scrollbar = tk.Scrollbar(self, orient="vertical", command=self.canvas.yview)
        self.scrollable_frame = tk.Frame(self.canvas, bg="#FFFFFF")

        self.scrollable_frame.bind(
            "<Configure>",
            lambda e: self.canvas.configure(
                scrollregion=self.canvas.bbox("all")
            )
        )

        self.canvas_frame = self.canvas.create_window((0, 0), window=self.scrollable_frame, anchor="nw")

        self.canvas.configure(yscrollcommand=self.scrollbar.set)
        self.canvas.pack(side="left", fill="both", expand=True)
        self.scrollbar.pack(side="left", fill="y")
        self.canvas.bind('<Configure>', self.FrameWidth)

    def FrameWidth(self, event):
        canvas_width = event.width
        self.canvas.itemconfig(self.canvas_frame, width=canvas_width)



functions.py
import os
import pickle
import requests as reqs
from datetime import date
import json
from start_view_section import StartViewSection

MIN_VOTES_MOVIE = ""
MIN_VOTES_TV = ""
API_KEY = ""
INCOGNITO = ""
BG_COLOR = ""
FONT = ""
FONT_COLOR = ""
SECOND_COLOR = ""
THIRD_COLOR = ""
ICONS_PATH = "%s/Icons/" % os.curdir

ACTUAL_HISTORY_OBJECT = None
ACTUAL_HISTORY_OBJECT_MASTER = None

"""
sends request and get results from all pages

:param what: type of production - movie or tv
:return: list of productions with min x votes
"""


def get_productions(what):
    if what == "movie":
        url = "https://api.themoviedb.org/3/discover/movie?api_key=" + API_KEY + \
              "&language=en-US&include_adult=false&include_video=false&vote_count.gte=" + MIN_VOTES_MOVIE
    elif what == "tv":
        url = "https://api.themoviedb.org/3/discover/tv?api_key=" + API_KEY + \
              "&language=en-US&include_null_first_air_dates=false&vote_count.gte=" + MIN_VOTES_TV

    response = reqs.get(url)
    response_dict = json.loads(response.text)
    all_productions = []
    for page in range(1, response_dict["total_pages"] + 1):

        url_page = url + "&page=" + str(page)
        response_page = reqs.get(url_page)
        response_dict_page = json.loads(response_page.text)
        for result in response_dict_page["results"]:
            all_productions.append(result)

    return all_productions


"""
sends request and get info about genres

:param what: type of production - movie or tv
:return: list of available genres and dictionary of mapped ids to genres
"""


def get_genres(what):
    if what == 1:
        what = "movie"
    elif what == 2:
        what = "tv"
    url = "https://api.themoviedb.org/3/genre/" + what + "/list?api_key=" + API_KEY + "&language=en-US"

    response = reqs.get(url)
    resp = json.loads(response.text)
    genres = []
    for i in resp["genres"]:
        genres.append(i["name"])

    return genres, resp


"""
from all productions chooses list of those that match user's filers

:param what: type of production - movie or tv
:param title: demanded title or it's part
:param genres: list of demanded genres
:param year_from: the earliest demanded year of release
:param year_to: the latest demanded year of release
:param vote_from: the smallest demanded average vote
:param vote_to: the biggest demanded average vote
:return: list of productions that matches the filters
"""


def filter_results(what, title, genres, year_from, year_to, vote_from, vote_to):
    results = get_productions(what)
    index_delete = []

    if what == "movie":
        prod_title = "title"
        genres_map = get_genres(1)[1]["genres"]
        date = "release_date"
    elif what == "tv":
        prod_title = "name"
        genres_map = get_genres(2)[1]["genres"]
        date = "first_air_date"

    for i in range(len(results)):

        if len(title) > 0:
            if title.lower() not in results[i][prod_title].lower():
                index_delete.append(i)
                continue

        if len(genres) > 0:
            to_delete = True

            for genre in genres:
                gen_id = 0
                for gen in genres_map:
                    if gen["name"] == genre:
                        gen_id = gen["id"]
                if gen_id in results[i]["genre_ids"]:
                    to_delete = False
                    break
                if not to_delete:
                    break
            if to_delete:
                index_delete.append(i)
                continue

        if len(year_from) > 0:
            if int(year_from) > int(results[i][date][:4]):
                index_delete.append(i)
                continue

        if len(year_to) > 0:
            if int(results[i][date][:4]) > int(year_to):
                index_delete.append(i)
                continue

        if len(vote_from) > 0:
            if float(vote_from) > float(results[i]["vote_average"]):
                index_delete.append(i)
                continue

        if len(vote_to) > 0:
            if float(results[i]["vote_average"]) > float(vote_to):
                index_delete.append(i)
                continue
    correct = []
    for i in range(len(results)):
        if i not in index_delete:
            correct.append(results[i])

    return correct


"""
sorts list of productions

:param what: type of production - movie or tv
:param results: list of productions to sort
:param sorting_type: demanded type of sorting
:return: sorted list of productions
"""


def sort_results(what, results, sorting_type):
    def switch_case(value):
        switch = {
            "A-Z": lambda: (["title", "name"], False),
            "Z-A": lambda: (["title", "name"], True),
            "release date - asc": lambda: (["release_date", "first_air_date"], False),
            "release date - desc": lambda: (["release_date", "first_air_date"], True),
            "popularity - asc": lambda: (["popularity", "popularity"], False),
            "popularity - desc": lambda: (["popularity", "popularity"], True),
            "vote average - asc": lambda: (["vote_average", "vote_average"], False),
            "vote average - desc": lambda: (["vote_average", "vote_average"], True),
        }
        func = switch.get(value, lambda: (["", ""], False))

        return func()

    if what == "movie":
        index = 0
    elif what == "tv":
        index = 1

    if len(sorting_type) > 0:
        results = sorted(results, key=lambda k: (k[switch_case(sorting_type)[0][index]]),
                         reverse=switch_case(sorting_type)[1])

    return results


"""
uses functions to filter and sort results

:param what: type of production - movie or tv
:param title: demanded title or it's part
:param genres: list of demanded genres
:param year_from: the earliest demanded year of release
:param year_to: the latest demanded year of release
:param vote_from: the smallest demanded average vote
:param vote_to: the biggest demanded average vote
:param sorting_type: demanded type of sorting
:return: filtered and sorted list of productions
"""


def search_filter_sort(what, title, genres, year_from, year_to, vote_from, vote_to, sorting_type):
    after_filtres = filter_results(what, title, genres, year_from, year_to, vote_from, vote_to)

    if len(sorting_type) > 0:
        after_filtres = sort_results(what, after_filtres, sorting_type)

    return after_filtres


"""
sends request about specific production

:param what: type of production - movie or tv
:param id: id of production
:param extended: boolean, chooses extended version of info
:return: list of information about demanded production
"""


def get_info_about(what, id, extended):
    url = "https://api.themoviedb.org/3/" + what + "/" + str(id) + "?api_key=" + API_KEY + "&language=en-US"
    response = reqs.get(url)
    resp = json.loads(response.text)
    genres = []
    for i in resp["genres"]:
        genres.append(i["name"])

    if extended:
        if what == "movie":
            info = [resp["title"], resp["release_date"][:4], genres, resp["poster_path"], resp["overview"],
                    resp["vote_average"]]
        elif what == "tv":
            info = [resp["name"], resp["first_air_date"][:4], resp["last_air_date"][:4], genres, resp["poster_path"],
                    resp["overview"], resp["status"], resp["vote_average"]]
    else:
        if what == "movie":
            info = ["movie", resp["id"], resp["poster_path"], resp["title"], resp["vote_average"]]
        elif what == "tv":
            info = ["tv", resp["id"], resp["poster_path"], resp["name"], resp["vote_average"]]

    return info


"""
chooses most important info about productions

:param what: type of production - movie or tv
:param data: dictionary with full info of productions
:return: list with most important info about productions
"""


def filter_info(what, data):
    info = []
    for prod in data:
        if what == "movie":
            info.append(["movie", prod["id"], prod["poster_path"], prod["title"], prod["vote_average"]])
        elif what == "tv":
            info.append(["tv", prod["id"], prod["poster_path"], prod["name"], prod["vote_average"]])

    return info


"""
sends request and check the oldest production within type

:param what: type of production - movie or tv
:return: year of the oldest production
"""


def get_oldest_release_year(what):
    if what == "movie":
        url = "https://api.themoviedb.org/3/discover/movie?api_key=" + API_KEY + \
              "&language=en-US&sort_by=primary_release_date.asc&include_adult=false&include_video=false&page=1&vote_count.gte=" + MIN_VOTES_MOVIE
    elif what == "tv":
        url = "https://api.themoviedb.org/3/discover/tv?api_key=" + API_KEY + \
              "&language=en-US&sort_by=first_air_date.asc&page=1&include_null_first_air_dates=false&vote_count.gte=" + MIN_VOTES_TV

    response = reqs.get(url)
    response_dict = json.loads(response.text)

    if what == "movie":
        return response_dict["results"][0]["release_date"][:4]
    elif what == "tv":
        return response_dict["results"][0]["first_air_date"][:4]


"""
chooses most important info about top productions from specific year

:param what: type of production - movie or tv
:param year: demanded year of release
:param how_many: demanded number of productions
:param reverse: boolean, chooses reverse order of results
:return: list with most important info about productions
"""


def get_top_year(what, year, how_many, reverse):
    productions = get_productions(what)
    correct = []

    for prod in productions:
        if what == "movie":
            if prod["release_date"][:4] == str(year):
                correct.append(prod)
        elif what == "tv":
            if prod["first_air_date"][:4] == str(year):
                correct.append(prod)

    correct = sorted(correct, key=lambda k: (k["vote_average"], k["popularity"]), reverse=reverse)[:how_many]

    info = []
    for prod in correct:
        if what == "movie":
            info.append(["movie", prod["id"], prod["poster_path"], prod["title"], prod["vote_average"]])
        elif what == "tv":
            info.append(["tv", prod["id"], prod["poster_path"], prod["name"], prod["vote_average"]])

    return info


"""
create file with most important info about top productions from specific year if required 
or read that info from existing file

:param what: type of production - movie or tv
:param year: demanded year of release
:param how_many: demanded number of productions
:param reverse: boolean, chooses reverse order of results
:return: list with most important info about productions
"""


def get_top_year_config_file(what, year, how_many, reverse):

    def save_to_file(name, object):
        with open(name, 'wb') as file:
            pickle.dump(object, file)

    def load_from_file(name):
        with open(name, 'rb') as file:
            return pickle.load(file)

    today = str(date.today())
    name = "config_" + what + "_" + today + ".txt"
    correct = ""
    for file_name in os.listdir(os.getcwd()):
        if file_name.startswith("config_" + what):
            correct = file_name

    if len(correct) == 0:
        info = get_top_year(what, year, how_many, reverse)
        save_to_file(name, info)
    else:
        if correct != name:
            os.remove(correct)
            info = get_top_year(what, year, how_many, reverse)
            save_to_file(name, info)
            return info

        info = load_from_file(name)

    return info


"""
add new production to history file and refresh displayed history

:param what: type of production - movie or tv
:param id: id of production
"""


def write_history(what, id):
    global ACTUAL_HISTORY_OBJECT, ACTUAL_HISTORY_OBJECT_MASTER
    if not INCOGNITO:
        lines = ""
        my_line = what + " " + str(id) + "\n"
        if os.path.exists("history.txt"):
            with open("history.txt", "r") as file:
                for line in file.readlines():
                    if line != my_line:
                        lines += line
        else:
            ACTUAL_HISTORY_OBJECT_MASTER.grid(column=0, row=2, pady=30, sticky='w', padx=30)
        lines += my_line
        with open("history.txt", "w") as file:
            file.write(lines)

        if ACTUAL_HISTORY_OBJECT is not None:
            ACTUAL_HISTORY_OBJECT = ""

        ACTUAL_HISTORY_OBJECT = StartViewSection(ACTUAL_HISTORY_OBJECT_MASTER, "last")


"""
reads file with history and takes max 10 elements

:return: list with info about productions from history and number how many to display
"""


def read_history():
    info = []
    with open("history.txt", "r") as file:
        lines = []
        for line in file.readlines():
            words = line.split(" ")
            words_to_add = []
            for word in words:
                words_to_add.append(word.replace("\n", ""))
            lines.append(words_to_add)
        if len(lines) != 0 and len(lines) <= 10:
            lines.reverse()
            how_many_last = len(lines)
        else:
            lines = lines[-10:]
            lines.reverse()
            how_many_last = 10

        for line in lines:
            details = get_info_about(line[0], line[1], True)
            if line[0] == "movie":
                poster = details[3]
            elif line[0] == "tv":
                poster = details[4]
            info.append([line[0], line[1], poster, details[0], details[-1]])

        return info, how_many_last


"""
deletes file with history 
"""

def delete_history():
    if os.path.exists("history.txt"):

        for widget in ACTUAL_HISTORY_OBJECT_MASTER.winfo_children():
            widget.grid_forget()
        ACTUAL_HISTORY_OBJECT_MASTER.grid_forget()
        os.remove("history.txt")
        global ACTUAL_HISTORY_OBJECT
        ACTUAL_HISTORY_OBJECT = None


"""
creates file with configuration info if required 
or read that info from existing file

:param name: name of config file
"""


def read_config_file(name):
    config_dict = { "MIN_VOTES_MOVIE": str(2500),
                    "MIN_VOTES_TV": str(500),
                    "API_KEY": "e4e6acdc6a73d9fbd3d5033d86283840",
                    "INCOGNITO": False,
                    "BG_COLOR": "#FFFFFF",
                    "FONT": "Century Gothic",
                    "FONT_COLOR": "#161616",
                    "SECOND_COLOR": "#FF2D55",
                    "THIRD_COLOR": "#008898"}

    if os.path.exists(name):
        with open(name, 'r') as file:
            for line in file.readlines():
                words = line.split("=")
                config_dict[words[0]] = words[1].replace("\n", "")
    else:
        with open(name, 'w') as file:
            for key, value in config_dict.items():
                file.write(str(key) + "=" + str(value) + "\n")

    global MIN_VOTES_MOVIE, MIN_VOTES_TV, API_KEY, INCOGNITO, BG_COLOR, FONT, FONT_COLOR, SECOND_COLOR, THIRD_COLOR, ICONS_PATH
    MIN_VOTES_MOVIE = config_dict["MIN_VOTES_MOVIE"]
    MIN_VOTES_TV = config_dict["MIN_VOTES_TV"]
    API_KEY = config_dict["API_KEY"]
    INCOGNITO = True if config_dict["INCOGNITO"] == "True" else False
    BG_COLOR = config_dict["BG_COLOR"]
    FONT = config_dict["FONT"]
    FONT_COLOR = config_dict["FONT_COLOR"]
    SECOND_COLOR = config_dict["SECOND_COLOR"]
    THIRD_COLOR = config_dict["THIRD_COLOR"]


start_view.py
import os
import tkinter as tk
import functions
from scroll import ScrollableFrame
from start_view_section import StartViewSection


class StartView:

    """
    creates start view with scroll, top movies, tv series from last year, and history
    """

    def __init__(self, master):
        self.scroll_frame = ScrollableFrame(master)
        self.scroll_frame.pack(fill='both', expand=True)

        self.master = master

        self.movies_frame = tk.LabelFrame(self.scroll_frame.scrollable_frame, bg=functions.BG_COLOR)
        self.tv_frame = tk.LabelFrame(self.scroll_frame.scrollable_frame, bg=functions.BG_COLOR)
        self.last_frame = tk.LabelFrame(self.scroll_frame.scrollable_frame, bg=functions.BG_COLOR)
        functions.ACTUAL_HISTORY_OBJECT_MASTER = self.last_frame

        self.movies_frame.grid(column=0, row=0, padx=30, pady=30)
        self.tv_frame.grid(column=0, row=1)
        self.last_frame.grid(column=0, row=2, pady=30, sticky='w', padx=30)

        StartViewSection(self.movies_frame, "movie")
        StartViewSection(self.tv_frame, "tv")
        if os.path.exists("history.txt"):
            history = StartViewSection(self.last_frame, "last")
            functions.ACTUAL_HISTORY_OBJECT = history



details_view.py
import tkinter as tk
from io import BytesIO
import requests as reqs
from PIL import ImageTk, Image
import functions
from scroll import ScrollableFrame


class DetailsView:

    """
    creates details view with info about specific production
    """

    def __init__(self, master, what, id, photo):

        self.scroll_frame = ScrollableFrame(master)
        self.scroll_frame.pack(fill='both', expand=True)

        self.details_frame = tk.LabelFrame(self.scroll_frame.scrollable_frame, bg=functions.BG_COLOR)
        self.details_frame.grid(column=0, row=0, padx=30, pady=50)

        self.info = functions.get_info_about(what, id, True)

        functions.write_history(what, id)

        self.place_poster(photo)
        self.place_info(self.info, what)

    """
    send request to get poster of production and places it
    """

    def place_poster(self, photo):
        if photo is not None:
            img_url = "https://image.tmdb.org/t/p/w300" + photo
            img_data = reqs.get(img_url).content
            img = ImageTk.PhotoImage(Image.open(BytesIO(img_data)).resize((300, 450), Image.ANTIALIAS))
        else:
            img_path = functions.ICONS_PATH + "300x450.png"
            img = tk.PhotoImage(file=img_path)

        photo_label = tk.Label(self.details_frame, image=img, bg="#FFFFFF")
        photo_label.image = img
        photo_label.grid(row=0, column=0, rowspan=5, sticky='nw')

    """
    places info about production
    """

    def place_info(self, info, what):

        title = tk.Message(self.details_frame, text=info[0], padx=10, font=(functions.FONT, 36), bg=functions.BG_COLOR,
                           fg=functions.FONT_COLOR, width=700)
        title.grid(row=0, column=1, sticky='w')

        if what == "movie":
            date = info[1]
            details_text = info[4]
            genres = info[2]
        elif what == "tv":
            if info[6] == "Ended":
                if info[1] == info[2]:
                    date = info[1]
                else:
                    date = str(info[1]) + " - " + str(info[2])
            else:
                date = str(info[1]) + " - now"

            details_text = info[5]
            genres = info[3]

        genres = ", ".join(genres)
        genres.replace("{", "").replace("}", "")

        genres_label = tk.Message(self.details_frame, text=genres, padx=10, font=(functions.FONT, 24), bg=functions.BG_COLOR, fg=functions.FONT_COLOR, width=700)
        genres_label.grid(row=1, column=1, sticky='nw')

        year = tk.Label(self.details_frame, text=date, padx=10, font=(functions.FONT, 24), bg=functions.BG_COLOR, fg=functions.FONT_COLOR)
        year.grid(row=2, column=1, sticky='nw')

        rating = tk.Label(self.details_frame, text=str(info[-1]) + "/10", padx=10, bg=functions.BG_COLOR, fg=functions.FONT_COLOR,
                          font=(functions.FONT, 18))
        rating.grid(row=3, column=1, sticky='w')

        details = tk.Message(self.details_frame, text=details_text, padx=10, font=(functions.FONT, 16), bg=functions.BG_COLOR,
                             fg=functions.FONT_COLOR, width=700)
        details.grid(row=4, column=1, sticky='w')

        self.details_frame.rowconfigure(0, weight=1)
        self.details_frame.rowconfigure(1, weight=1)
        self.details_frame.rowconfigure(2, weight=1)
        self.details_frame.rowconfigure(3, weight=1)
        self.details_frame.rowconfigure(4, weight=5)


single_prod_view.py
import tkinter as tk
from io import BytesIO
import requests as reqs
from PIL import ImageTk, Image
import functions
from details_view import DetailsView

"""
creates window with details of specific production
"""


def create_window_details(what, id, photo):
    top = tk.Toplevel()
    top.wm_geometry("1100x500")
    DetailsView(top, what, id, photo)


class SingleProdView:

    """
    creates single production view with basic info about it
    """

    def __init__(self, master, table, i, data_source, number):
        self.m_frame = tk.Frame(master, bg=functions.BG_COLOR)

        # NUMBERS - ORDER
        if number:
            self.number = tk.Label(self.m_frame, text=(i + 1), padx=10, font=(functions.FONT, 20), anchor='w',
                                   bg=functions.BG_COLOR, fg=functions.FONT_COLOR)
            self.number.grid(row=0, sticky='w')

        # POSTER
        if data_source[i][2] is not None:
            img_url = "https://image.tmdb.org/t/p/w200" + data_source[i][2]
            img_data = reqs.get(img_url).content
            img = ImageTk.PhotoImage(Image.open(BytesIO(img_data)).resize((200, 300), Image.ANTIALIAS))
        else:
            img_path = functions.ICONS_PATH + "200x300.png"
            img = tk.PhotoImage(file=img_path)

        self.photo_label = tk.Label(self.m_frame, image=img, bg=functions.BG_COLOR)
        self.photo_label.image = img
        self.photo_label.grid(row=1)
        self.m_frame.rowconfigure(1, minsize=320)
        self.m_frame.rowconfigure(2, minsize=80)
        self.m_frame.rowconfigure(3, minsize=50)

        self.show_more_button = tk.Button(self.m_frame, text="Show details",
                                          command=lambda i_value=i: create_window_details(data_source[i_value][0],
                                                                                          data_source[i_value][1],
                                                                                          data_source[i_value][2]),
                                          fg="#FFFFFF",
                                          font=(functions.FONT, 11), bg=functions.SECOND_COLOR)
        self.show_more_button.grid(row=3, sticky='ne')

        self.rating = tk.Label(self.m_frame, text=str(data_source[i][-1]) + "/10", padx=5, bg=functions.BG_COLOR, fg=functions.FONT_COLOR,
                               font=(functions.FONT, 14))
        self.rating.grid(row=3, sticky='nw', pady=3)

        self.title_message = tk.Message(self.m_frame, text=data_source[i][3], padx=5, width=200, bg=functions.BG_COLOR,
                                        fg=functions.FONT_COLOR,
                                        font=(functions.FONT, 14))
        self.title_message.grid(row=2, sticky='nw')
        self.m_frame.update()

        table.append(self.m_frame)


start_view_section.py
import os
import tkinter as tk
from datetime import date
import functions
from single_prod_view import SingleProdView


class StartViewSection:

    """
    creates start view section with single productions and horizontal scroll in buttons
    """

    def __init__(self, master, what):

        self.master = master
        self.what = what

        self.index = 0
        self.how_many = 10

        self.last_year = date.today().year - 1

        self.table = []
        right_path = functions.ICONS_PATH + "next.png"
        left_path = functions.ICONS_PATH + "back.png"
        self.right_arrow = tk.PhotoImage(file=right_path)
        self.left_arrow = tk.PhotoImage(file=left_path)

        self.create_top(what)
        self.show_top(what)

    """
    creates views of single productions
    
    :param what: type of production - movie or tv
    """

    def create_top(self, what):

        if what == "movie":
            top_10 = functions.get_top_year_config_file("movie", self.last_year, self.how_many, True)
        elif what == "tv":
            top_10 = functions.get_top_year_config_file("tv", self.last_year, self.how_many, True)
        elif what == "last":
            if os.path.exists("history.txt"):
                self.info, self.how_many = functions.read_history()
            top_10 = self.info
            self.table.clear()

        for i in range(len(top_10)):
            SingleProdView(self.master, self.table, i, top_10, True)

    """
    shows views of single productions with scroll buttons if required (number>5)

    :param what: type of production - movie or tv
    """

    def show_top(self, what):

        if what == "movie":
            text = str(self.last_year) + "'s top 10 movies"
        elif what == "tv":
            text = str(self.last_year) + "'s top TV series"
        elif what == "last":
            text = "last viewed productions"

        text = tk.Label(self.master, text=text, padx=10, font=(functions.FONT, 26), bg=functions.BG_COLOR, fg=functions.FONT_COLOR)
        text.grid(row=0, columnspan=3, sticky='w')

        # LEFT BUTTON
        if len(self.table) > 5:
            left_button = tk.Button(self.master, image=self.left_arrow, command=lambda: self.left_click(what))
            left_button.grid(column=0, row=1)
            self.master.grid_columnconfigure(0, minsize=50)
            self.master.grid_columnconfigure(6, minsize=50)

        for item in self.table:
            item.grid_forget()

        counter = 0
        if len(self.table) > 5:
            my_range = self.index + 5
        else:
            my_range = len(self.table)
        for i in range(self.index, my_range):
            counter += 1
            self.table[i].grid(column=counter, row=1, padx=10, sticky='n')

        # RIGHT BUTTON
        if len(self.table) > 5:
            right_button = tk.Button(self.master, image=self.right_arrow, command=lambda: self.right_click(what))
            right_button.grid(column=6, row=1)

    """
    scroll left by one item
    """

    def left_click(self, what):
        if self.index >= 1:
            self.index -= 1

        self.show_top(what)

    """
    scroll right by one item
    """

    def right_click(self, what):
        if self.index <= self.how_many - 6:
            self.index += 1

        self.show_top(what)

main_app.py
import tkinter as tk
import webbrowser
from tkinter import ttk
from tkinter import messagebox
import functions
from search_view import SearchView
from start_view import StartView

"""
opens window with help
"""


def help_view():
    top = tk.Toplevel(bg=functions.BG_COLOR)

    top.wm_geometry("750x350")
    info = tk.Message(top, text="""Welcome to our app!
    
We're here to help you find any movie or tv series. Just look, it's very simple.

On the home page you can find best productions of last year and your history. You can switch the incognito mode if you don't want to save your history. There is a button in a bottom part. 

If you want to personalize your query just go to search page, choose whatever you're interested in and click 'SEARCH' button

Hope you'll enjoy it!
""", font=(functions.FONT, 12), width=700, pady=50, bg=functions.BG_COLOR, fg=functions.FONT_COLOR)
    info.pack()


"""
opens site of TMDb
"""


def open_site():
    webbrowser.open('https://www.themoviedb.org/')


class BaseWindow:

    """
    creates base window with menu, bottombar and tabs
    """

    def __init__(self, master):
        self.master = master
        self.master.title("TV Maniac")

        # MENU
        self.menubar = tk.Menu(self.master)
        self.master["menu"] = self.menubar

        self.file_menu = tk.Menu(self.menubar)
        self.file_menu.add_command(label="Exit", underline=0, command=self.exit)

        self.history_menu = tk.Menu(self.menubar)
        self.history_menu.add_command(label="Delete history", underline=0, command=functions.delete_history)
        self.history_menu.add_command(label="Switch incognito mode", underline=0, command=self.switch_incognito)

        self.help_menu = tk.Menu(self.menubar)
        self.help_menu.add_command(label="TMDb's site", underline=0, command=open_site)
        self.help_menu.add_command(label="Help", underline=0, command=help_view)

        self.menubar.add_cascade(label="File", menu=self.file_menu, underline=0)
        self.menubar.add_cascade(label="History", menu=self.history_menu, underline=0)
        self.menubar.add_cascade(label="About", menu=self.help_menu, underline=0)

        # BOTTOMBAR
        self.bottombar = tk.LabelFrame(master)
        self.bottombar.pack(side=tk.BOTTOM, anchor='w', fill='x')
        logo_path = functions.ICONS_PATH + "logo.png"
        self.logo = tk.PhotoImage(file=logo_path)
        self.logo_label = tk.Label(self.bottombar, image=self.logo)
        self.logo_label.pack(side=tk.LEFT, padx=20, pady=5)
        self.statusbar_text = tk.Label(self.bottombar,
                                       text="This product uses the TMDb API but is not endorsed or certified by TMDb",
                                       font=(functions.FONT, 10))
        self.statusbar_text.pack(side=tk.LEFT)

        self.incognito_button = tk.Button(self.bottombar, fg="#FFFFFF", command=self.switch_incognito, text="OFF",
                                          bg="#8B0000", width=10)
        self.incognito_button.pack(side=tk.RIGHT, padx=20)

        self.incognito_text = tk.Label(self.bottombar, text="INCOGNITO MODE:", font=(functions.FONT, 11))
        self.incognito_text.pack(side=tk.RIGHT)

        # TABS
        self.tab_control = ttk.Notebook(self.master)
        self.tab_control.pack(expand=1, fill="both")

        self.home_tab = tk.Frame(self.tab_control)
        self.start = StartView(self.home_tab)

        self.search_tab = tk.Frame(self.tab_control)
        self.search = SearchView(self.search_tab)

        self.tab_control.add(self.home_tab, text='Home')
        self.tab_control.add(self.search_tab, text='Search')

    """
    exits from app
    """

    def exit(self):
        answer = tk.messagebox.askyesno("TVManiac", "Do you want to exit?", parent=self.master)
        if answer:
            self.master.destroy()

    """
    switches incognito mode
    """

    def switch_incognito(self):
        if functions.INCOGNITO:
            self.incognito_button.configure(bg="#8B0000", text="OFF")
            functions.INCOGNITO = False
        else:
            self.incognito_button.configure(bg="#006400", text="ON")
            functions.INCOGNITO = True


if __name__ == '__main__':
    root = tk.Tk()
    root.geometry('1300x750')
    functions.read_config_file("config_file.txt")
    icon_path = functions.ICONS_PATH + "television.png"
    root.iconphoto(True, tk.PhotoImage(file=icon_path))
    root.resizable(False, False)
    app = BaseWindow(root)
    root.mainloop()
